docs/release.md: Add release documentation (#1712)
libp2p/rust-libp2p: docs/release.md

# Release Process

This project follows [semantic versioning](https://semver.org/). The following
documentation will refer to `X.Y.Z` as _major_, _minor_ and _patch_ version.

> ## Before the Release

Reviewer: ```suggestion
## Development between releases
```

- Every substantial pull request should add an entry to the `[unreleased]`
  section of the corresponding crate `CHANGELOG.md` file. See
  [#1698](https://github.com/libp2p/rust-libp2p/pull/1698/files) as an example.
  
  In case there is no `[unreleased]` section yet, create one with an increased
  major, minor or patch version depending on your change. In addition update the
  corresponding entry of the crate in the root level `Cargo.toml`.

> ## Cutting a Release

Reviewer: ```suggestion
## Releasing one or more crates
```

> 1. For each crate's `CHANGELOG.md` including the top level `libp2p` crate
>    replace `# X.Y.Z [unreleased]` with `# X.Y.Z [yyyy-mm-dd]` and create a
>    combined pull request against the rust-libp2p `master` branch.

Reviewer: ```suggestion
1. Set the release date for each crate to be released in the respective `CHANGELOG.md`
   by replacing `# X.Y.Z [unreleased]` with `# X.Y.Z [yyyy-mm-dd]` and create a pull request
   against the rust-libp2p `master` branch.
```

   
2. Once merged, create and push a tag for each updated crate.

    ```
    $ tag="<CRATE-NAME>-X.Y.Z"
    $ git tag -s "${tag}" -m "${tag}"
    $ git push origin "${tag}"
    ```
    

> 3. Create and push a tag for the top level `libp2p` crate.

Reviewer: ```suggestion
3. Create and push a tag for the top level `libp2p` crate, if it is being released.
```

    ```
    # Note the additional `v` here.
    $ tag="vX.Y.Z"
    $ git tag -s "${tag}" -m "${tag}"
    $ git push origin "${tag}"
    ```
    

> 4. Publish each crate including the top level `libp2p` crate to crates.io.

Reviewer: ```suggestion
4. Publish each tagged crate to crates.io. `cargo` assists in getting the order of the releases correct.
```

# Release process

This project follows [semantic versioning](https://semver.org/). The following
documentation will refer to `X.Y.Z` as _major_, _minor_ and _patch_ version.

## Development between releases

- Every substantial pull request should add an entry to the `[unreleased]`
  section of the corresponding crate `CHANGELOG.md` file. See
  [#1698](https://github.com/libp2p/rust-libp2p/pull/1698/files) as an example.
  
  In case there is no `[unreleased]` section yet, create one with an increased
  major, minor or patch version depending on your change. In addition update the
  corresponding entry of the crate in the root level `Cargo.toml`.

## Releasing one or more crates

1. Set the release date for each crate to be released in the respective
   `CHANGELOG.md` by replacing `# X.Y.Z [unreleased]` with

>    `# X.Y.Z [yyyy-mm-dd]` and create a pull request against the rust-libp2p

Reviewer 2: I wonder why we insist on putting those dates in there? The version corresponds to a tag which maps to a commit which has a date. How often do people look at the date in a changelog?

Author: I am fine either way. I just want to be consistent across all `github.com/libp2p/rust-libp2p` crates. @romanb thoughts?

Reviewer: I don't care about the dates in the changelogs much. That tradition was started [here](https://github.com/libp2p/rust-libp2p/pull/1614#discussion_r442206372).

   `master` branch.

2. Once merged, create and push a tag for each updated crate.

    ```
    $ tag="<CRATE-NAME>-X.Y.Z"

>     $ git tag -s "${tag}" -m "${tag}"

Reviewer 2: I have never created a GPG signed tag so far, but went for annotated tags. I also create them all locally and eventually push them altogether  with `git push origin master --tags`.

Author: Added a comment above. Let me know if that looks good to you @twittner.
